pkg/smokescreen/smokescreen.go

```diff
@@ -433,11 +433,25 @@ func runServer(config *Config, server *http.Server, listener net.Listener, quit
 		config.Log.Fatal(err)
 	}
 
-	if semiGraceful {
-		// the program has exited normally, wait 60s in an attempt to shutdown
-		// semi-gracefully
-		config.Log.WithField("delay", config.ExitTimeout).Info("Waiting before shutting down")
-		time.Sleep(config.ExitTimeout)
+	if graceful {
+		// the program has exited normally, wait for all connections to become idle before
+		// continuing in an attempt to shutdown gracefully
+		beginTs := time.Now()
+		alerted := false
+		for {
+			if config.StatsServer.(*StatsServer).GetNumActiveConn() > 0 {
+				config.Log.Info(fmt.Sprintf("There are still active connections. Waiting %v before checking again.", config.WaitForAllIdleSec))
+				if time.Now().Sub(beginTs) > config.ExitTimeout && !alerted {
+					config.Log.Info(fmt.Sprintf("We've been waiting for %v to shut down but there are still active connections.", config.ExitTimeout))
+					// Do alerting here
```

Comment:
I want to send a Slack alert or Sentry alert here...is that currently possible in this repo?

Reply:
Not currently.  You could do something similar to `Config.RoleFromRequest` where there's a default behavior wired up in Smokescreen's `main()`, but another program that pulls in the `smokescreen` package can provide environment-specific behavior.  (Alternately, you could send a metric to statsd, and trigger an alert on the far side of that?)